[김주형] Mission 3: 회원가입 기능 구현

BnDC opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 회원가입 기능 구현
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 이메일, 비밀번호 입력 검사 분리
- 정규표현식을 위한 불필요한 변수 제거 및 조건문에 직접 작성
- profile null값 허용으로 model 수정 및 makemigration + migrate
- profile 값 없으면 Null, 있으면 받게 수정
- slash 위치 변경

## 기타 질문 및 특이 사항


## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [X] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [X] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [X] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [X] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

BnDC committed:
[Add] SignUpView로 회원가입 구현
- 이메일, 비밀번호 입력되었는지 검사
- 이메일 중복 검사
- 이메일, 비밀번호 조건 충족 여부 검사

BnDC committed:
[Add] users.app URL 연결

BnDC committed:
[Add] users app makemigration + migrate

BnDC committed:
[Modified] Views.py 현재 필요없는 Module import,부분 제거

BnDC committed:
[Modified] Signupsview post error 수정
- re.error: password_checker (?=.*)에서 .누락 (두 시간 순삭)

BnDC committed:
[Modified] user 변수 변경

kyeongraekim commented on students/juhyeong/users/views.py:
```diff
@@ -1,3 +1,35 @@
+        if user.filter(email=data['email']).exists():
+            return JsonResponse({"message": "이메일이 이미 등록되었습니다"}, status =401)
+
+        elif not email_checker.match(data['email']) != None:
+            return JsonResponse({"message": "이메일 형식이 잘못 되었습니다"}, status =401)
+        
+        elif len(data['password']) < 8 or not password_checker.match(data['password']) != None:
```
- if문으로 논리전개를 펼칠 수 있고, elif 문을 굳이 사용할 필요가 없다면 사용하지 않아도 됩니다.

- "8" 이라는 숫자를 직접 작성하는 것보다  "PASSWORD_LENGTH = 8"로 숫자를 변수/상수 처리를 해서 관리하는 것이 추후에 비밀번호 최소 숫자(기획변경) 시 수정에 용이합니다. 
하지만 그보다 앞서서 정규표현식을 잘 써주셨는데 정규표현식을 활용해서 비밀번호 최소 길이를 설정해 놓을 수 있습니다. 정규표현식을 활용하여 수정해보세요!

BnDC replied:
불필요한 변수 선언, 필요 없는 조건(`len(data['password']) < 8`) 제거하고, 조건문에 정규 표현식 직접 작성하도록 하겠습니다.

kyeongraekim commented on students/juhyeong/users/views.py:
```diff
@@ -1,3 +1,35 @@
+        data             = json.loads(request.body)
+        email_checker    = re.compile('^[\w]+@[\w.\-]+\.[A-Za-z]{2,3}$')
+        password_checker = re.compile('^(?=.*[A-Za-z])(?=.*\d)(?=.*[~!@#$^&*()+|=])[A-Za-z\d~!@#$%^&*()+|=]{8,}$')
+        user             = User.objects
```
이 정도의 짧은 객체 호출 코드를 변수로 선언하는 것이 더 좋지 않아 보이네요. 아래 코드에 그냥 녹여주셔도 됩니다~

kyeongraekim commented on students/juhyeong/users/views.py:
```diff
@@ -1,3 +1,35 @@
+        password_checker = re.compile('^(?=.*[A-Za-z])(?=.*\d)(?=.*[~!@#$^&*()+|=])[A-Za-z\d~!@#$%^&*()+|=]{8,}$')
+        user             = User.objects
+
+        if 'email' not in data or 'password' not in data:
```
이메일과 패스워드를 하나의 조건문에서 모두 거르려고 한다면 정확한 에러 메세지를 전달할 수 없겠죠?
이메일이 입력이 되지 않아서 나온 에러인지, 패스워드가 에러가 나서 난 에러인지 최소한의 코드 분리를 통해서라도 알아 볼 수 있게 만들어주셔야 합니다.

BnDC replied:
![image](https://user-images.githubusercontent.com/86050295/135033087-ab856e4e-da38-429c-bb1b-d5d5ea6b9274.png)

질문드립니다 if 문을 분리하면 error message가 같더라도, 어디서 error 났는지 알 수 있는 방법이 있나요??

BnDC replied:
지금 if 문만 제거하고, 
email 없이, http post 요청하니  `KeyError: 'email'` 500
password없이, http post 요청하니 `KeyError: 'password'` 500
이 반환 됩니다.

이대로 if문만 제거해도 무방할까요?

kyeongraekim commented on students/juhyeong/users/views.py:
```diff
@@ -1,3 +1,35 @@
+            email        = data['email'],
+            password     = data['password'],
+            phone_number = data['phone_number'],
+            profile      = data['profile']
```
profile의 변수에 담길 데이터가 추가 정보라는 판단하에, 회원가입시 필수 입력정보가 아닌 추가 입력 정보 사항으로 기획 변경을 해보는 것은 어떨까요?

BnDC replied:
profile column을 null값이 가능하게 허용하게 모델 변경하고, user 생성시 profile 값을 필수적으로 받지 않도록 수정하겠습니다

kyeongraekim commented on students/juhyeong/westargram/urls.py:
```diff
@@ -13,7 +13,8 @@
+from django.urls import path, include
 
 urlpatterns = [
+    path('user/', include('users.urls'))
```
westargram/urls.py 에서 trailing slash를 붙이지 말고, users/urls.py 에서  slash를 앱 앞에 붙이는 것을 권장드립니다.

BnDC replied:
slash 방향 수정 하겠습니다

kyeongraekim reviewed (changes requested):
주형님~ 아래 리뷰사항 참고해서 수정 반영 부탁드려요~

BnDC committed:
[Modified] views.py
- email, password 입력 안할시 error 분리

BnDC committed:
[Modified] views.py
- 조건문에 정규표현식 직접 기입하도록 변경

BnDC committed:
[Modified] models.py
- profile null=True 추가

BnDC committed:
[Modified] makemigration + migrate 0002

BnDC committed:
[Modified] user 생성시 필수값은 받고, profile은 들어오면 받도록 수정

BnDC committed:
[Modified] views.py
불필요한 변수:email_checker, password_checker 제거

BnDC committed:
[modified] urls.py
 - path('user/', include('users.urls')) slash 위치 변경
 - path('/signup', SignUpsView.as_view())

lordmyshepherd commented:
LGTM 👍
